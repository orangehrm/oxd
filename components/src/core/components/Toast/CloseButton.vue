<template>
  <div class="oxd-toast-close-container">
    <div :class="classes" role="button">
      Ã—
    </div>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {toastMixin} from './toast-mixin';

export default defineComponent({
  name: 'oxd-toast-close-button',

  mixins: [toastMixin],

  computed: {
    classes(): object {
      return {
        'oxd-toast-close': true,
        [`oxd-toast-close--${this.type}`]: true,
      };
    },
  },
});
</script>

<style src="./close-button.scss" lang="scss" scoped></style>
