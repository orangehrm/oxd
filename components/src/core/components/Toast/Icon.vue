<!--
/*
 * This file is part of OrangeHRM Inc
 *
 * Copyright (C) 2020 onwards OrangeHRM Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see  http://www.gnu.org/licenses
 */
-->

<template>
  <div class="oxd-toast-icon-container">
    <div :class="iconWrapClasses">
      <oxd-icon :name="iconName" class="oxd-toast-icon" />
    </div>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {toastMixin} from './toast-mixin';
import {ToastType, ICON_MAP} from './types';
import Icon from '@ohrm/oxd/core/components/Icon/Icon.vue';

export default defineComponent({
  name: 'OxdToastIcon',

  components: {
    'oxd-icon': Icon,
  },

  mixins: [toastMixin],

  computed: {
    iconWrapClasses(): object {
      return {
        'oxd-toast-icon-wrap': true,
        [`oxd-toast-icon-wrap--${this.type}`]: true,
      };
    },
    iconName(): string {
      return ICON_MAP[this.type as ToastType];
    },
  },
});
</script>

<style src="./icon.scss" lang="scss" scoped></style>
