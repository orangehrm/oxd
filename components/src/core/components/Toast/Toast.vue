<template>
  <div v-if="show" :class="classes" aria-live="assertive" @click="onClickToast">
    <slot name="icon">
      <oxd-toast-icon :type="type" />
    </slot>
    <slot>
      <div :class="contentClasses">
        <oxd-text type="toast-title" class="oxd-toast-content-text">
          {{ title }}
        </oxd-text>
        <oxd-text type="toast-message" class="oxd-toast-content-text">
          {{ message }}
        </oxd-text>
      </div>
    </slot>
    <slot name="close">
      <oxd-toast-close-button :type="type" @click="onClickClose" />
    </slot>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {toastMixin} from './toast-mixin';
import Text from '@orangehrm/oxd/core/components/Text/Text.vue';
import ToastCloseButton from '@orangehrm/oxd/core/components/Toast/CloseButton.vue';
import ToastIcon from '@orangehrm/oxd/core/components/Toast/Icon.vue';

export default defineComponent({
  name: 'oxd-toast',

  mixins: [toastMixin],

  components: {
    'oxd-text': Text,
    'oxd-toast-close-button': ToastCloseButton,
    'oxd-toast-icon': ToastIcon,
  },

  props: {
    show: {
      type: Boolean,
      default: false,
    },
    title: {
      type: String,
      required: true,
    },
    message: {
      type: String,
    },
    persistent: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    classes(): object {
      return {
        'oxd-toast': true,
        [`oxd-toast--${this.type}`]: true,
        ['oxd-toast--persistent']: this.persistent,
      };
    },
    contentClasses(): object {
      return {
        'oxd-toast-content': true,
        [`oxd-toast-content--${this.type}`]: true,
      };
    },
  },

  methods: {
    onClickToast() {
      if (!this.persistent) {
        this.$emit('update:show', false);
      }
    },
    onClickClose(e: Event) {
      e.stopPropagation();
      this.$emit('update:show', false);
    },
  },
});
</script>

<style src="./toast.scss" lang="scss" scoped></style>
