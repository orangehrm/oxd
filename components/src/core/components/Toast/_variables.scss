$oxd-toast-icon-color-default: $oxd-interface-gray-darken-1-color !default;
$oxd-toast-icon-color: $oxd-white-color !default;
$oxd-toast-font-color-default: $oxd-interface-gray-darken-1-color !default;
$oxd-toast-font-color: $oxd-white-color !default;
$oxd-toast-border-radius: 3.2rem !default;
$oxd-toast-padding: 0.6rem !default;
$oxd-toast-box-shadow: 0px 25px 35px 0 rgba(0, 0, 0, 0.18) !default;
$oxd-toast-opacity: 0.8 !default;
$oxd-toast-content-padding-horizontal: 0.8rem !default;

$oxd-toast-bg-color-default: $oxd-white-color !default;
$oxd-toast-bg-color-success: $oxd-feedback-success-color !default;
$oxd-toast-bg-color-warn: $oxd-feedback-warn-color !default;
$oxd-toast-bg-color-error: $oxd-feedback-danger-color !default;
$oxd-toast-bg-color-info: $oxd-feedback-info-color !default;

$oxd-toast-icon-wrap-size: 3.2rem !default;
$oxd-toast-icon-wrap-padding: 1rem !default;
$oxd-toast-icon-font-size: 1.5rem !default;

$oxd-toast-close-size: 1rem !default;
$oxd-toast-close-padding: 0.2rem !default;
$oxd-toast-close-border-radius: 1rem !default;
