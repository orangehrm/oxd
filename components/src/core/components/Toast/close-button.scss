@import '../../../styles';
@import 'variables';

.oxd-toast-close {
  border: 0;
  outline: 0;
  padding: $oxd-toast-close-padding;
  height: $oxd-toast-close-size;
  width: $oxd-toast-close-size;
  border-radius: $oxd-toast-close-border-radius;
  cursor: pointer;
  background-color: rgba($oxd-toast-icon-color, 0.1);
  @include oxd-flex-center;

  &--default {
    background-color: rgba($oxd-toast-icon-color-default, 0.1);
  }
}

.oxd-toast-close-container {
  @include oxd-flex-center;
}
