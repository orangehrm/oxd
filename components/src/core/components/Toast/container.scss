@import '../../../styles';
@import 'variables';

.oxd-toast-container {
  display: flex;
  flex-direction: column-reverse;
  &--bottom {
    flex-direction: column-reverse;
  }
  &--top {
    flex-direction: column;
  }
}

.oxd-toast-container--toast {
  margin-bottom: $oxd-toast-container-space-between;
}

.oxd-toast-fade {
  // https://v3.vuejs.org/guide/transitions-enterleave.html#transitioning-single-elements-components
  &-enter-active {
    transition: opacity 0.5s ease-out;
  }
  &-enter-from {
    opacity: 0;
  }

  &-leave-active {
    transition: opacity 1s cubic-bezier(1, 0.5, 0.8, 1);
  }
  &-leave-to {
    opacity: 0 !important;
  }
}
