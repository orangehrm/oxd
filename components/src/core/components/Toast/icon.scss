@import '../../../styles';
@import 'variables';

.oxd-toast-icon {
  color: inherit;
  font-size: $oxd-toast-icon-font-size;
}

.oxd-toast-icon-container {
  @include oxd-flex-center;
}

.oxd-toast-icon-wrap {
  background-color: rgba($oxd-toast-icon-color, 0.1);
  padding: $oxd-toast-icon-wrap-padding;
  border-radius: 100%;
  box-sizing: border-box;
  height: $oxd-toast-icon-wrap-size;
  width: $oxd-toast-icon-wrap-size;
  @include oxd-flex-center;

  &--default {
    background-color: rgba($oxd-toast-icon-color-default, 0.1);
  }
}
