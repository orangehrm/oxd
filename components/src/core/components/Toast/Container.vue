<template>
  <div :class="classes">
    <transition-group :name="transition">
      <oxd-toast
        v-for="(toast, index) in toasts"
        :key="toast"
        v-bind="toast"
        @update:show="onUpdateShow($event, index, toast)"
        :class="toastClass"
        :style="{zIndex: toasts.length - index}"
      />
    </transition-group>
  </div>
</template>

<script lang="ts">
import {defineComponent} from 'vue';
import {POSITION_BOTTOM, POSITIONS, ToastContainerPositions} from './types';
import Toast from '@orangehrm/oxd/core/components/Toast/Toast.vue';

export default defineComponent({
  name: 'oxd-toast-container',

  components: {
    'oxd-toast': Toast,
  },

  emits: ['update:toasts'],

  props: {
    toasts: {
      type: Array,
      default: () => [],
    },
    transition: {
      type: String,
      default: 'oxd-toast-fade',
    },
    toastClass: {
      type: String,
      default: 'oxd-toast-container--toast',
    },
    position: {
      type: String,
      default: POSITION_BOTTOM,
      validator: function(value: ToastContainerPositions) {
        return POSITIONS.indexOf(value) !== -1;
      },
    },
  },

  computed: {
    classes(): object {
      return {
        'oxd-toast-container': true,
        [`oxd-toast-container--${this.position}`]: true,
      };
    },
  },

  methods: {
    onUpdateShow(state: boolean, index: number) {
      if (state === false) {
        const toasts = [...this.toasts];
        toasts.splice(index, 1);
        this.$emit('update:toasts', toasts);
      }
    },
  },
});
</script>

<style src="./container.scss" lang="scss" scoped></style>
