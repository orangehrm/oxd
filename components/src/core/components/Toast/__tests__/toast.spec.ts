import {mount} from '@vue/test-utils';
import Toast from '@orangehrm/oxd/core/components/Toast/Toast.vue';

describe('Toast > Toast.vue', () => {
  it('should renders OXD Toast', () => {
    const wrapper = mount(Toast, {
      props: {},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Toast success', () => {
    const wrapper = mount(Toast, {
      props: {type: 'success'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Toast warn', () => {
    const wrapper = mount(Toast, {
      props: {type: 'warn'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Toast error', () => {
    const wrapper = mount(Toast, {
      props: {type: 'error'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should renders OXD Toast info', () => {
    const wrapper = mount(Toast, {
      props: {type: 'info'},
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
