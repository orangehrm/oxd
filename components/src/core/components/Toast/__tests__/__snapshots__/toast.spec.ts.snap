// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toast > Toast.vue should renders OXD Toast 1`] = `
<div class="oxd-toast oxd-toast--default" aria-live="assertive">
  <div class="oxd-toast-start">
    <div class="oxd-toast-icon-container">
      <div class="oxd-toast-icon-wrap oxd-toast-icon-wrap--default"><i class="oxd-icon bi-star oxd-toast-icon"></i></div>
    </div>
    <div class="oxd-toast-content oxd-toast-content--default">
      <p class="oxd-text oxd-text--p oxd-text--toast-title oxd-toast-content-text">Test Toast</p>
      <p class="oxd-text oxd-text--p oxd-text--toast-message oxd-toast-content-text"></p>
    </div>
  </div>
  <div class="oxd-toast-close-container">
    <div class="oxd-toast-close oxd-toast-close--default" role="button"> × </div>
  </div>
</div>
`;

exports[`Toast > Toast.vue should renders OXD Toast error 1`] = `
<div class="oxd-toast oxd-toast--error" aria-live="assertive">
  <div class="oxd-toast-start">
    <div class="oxd-toast-icon-container">
      <div class="oxd-toast-icon-wrap oxd-toast-icon-wrap--error"><i class="oxd-icon bi-exclamation-circle oxd-toast-icon"></i></div>
    </div>
    <div class="oxd-toast-content oxd-toast-content--error">
      <p class="oxd-text oxd-text--p oxd-text--toast-title oxd-toast-content-text">Test Toast</p>
      <p class="oxd-text oxd-text--p oxd-text--toast-message oxd-toast-content-text"></p>
    </div>
  </div>
  <div class="oxd-toast-close-container">
    <div class="oxd-toast-close oxd-toast-close--error" role="button"> × </div>
  </div>
</div>
`;

exports[`Toast > Toast.vue should renders OXD Toast info 1`] = `
<div class="oxd-toast oxd-toast--info" aria-live="assertive">
  <div class="oxd-toast-start">
    <div class="oxd-toast-icon-container">
      <div class="oxd-toast-icon-wrap oxd-toast-icon-wrap--info"><i class="oxd-icon bi-info-circle oxd-toast-icon"></i></div>
    </div>
    <div class="oxd-toast-content oxd-toast-content--info">
      <p class="oxd-text oxd-text--p oxd-text--toast-title oxd-toast-content-text">Test Toast</p>
      <p class="oxd-text oxd-text--p oxd-text--toast-message oxd-toast-content-text"></p>
    </div>
  </div>
  <div class="oxd-toast-close-container">
    <div class="oxd-toast-close oxd-toast-close--info" role="button"> × </div>
  </div>
</div>
`;

exports[`Toast > Toast.vue should renders OXD Toast success 1`] = `
<div class="oxd-toast oxd-toast--success" aria-live="assertive">
  <div class="oxd-toast-start">
    <div class="oxd-toast-icon-container">
      <div class="oxd-toast-icon-wrap oxd-toast-icon-wrap--success"><i class="oxd-icon bi-check2 oxd-toast-icon"></i></div>
    </div>
    <div class="oxd-toast-content oxd-toast-content--success">
      <p class="oxd-text oxd-text--p oxd-text--toast-title oxd-toast-content-text">Test Toast</p>
      <p class="oxd-text oxd-text--p oxd-text--toast-message oxd-toast-content-text"></p>
    </div>
  </div>
  <div class="oxd-toast-close-container">
    <div class="oxd-toast-close oxd-toast-close--success" role="button"> × </div>
  </div>
</div>
`;

exports[`Toast > Toast.vue should renders OXD Toast warn 1`] = `
<div class="oxd-toast oxd-toast--warn" aria-live="assertive">
  <div class="oxd-toast-start">
    <div class="oxd-toast-icon-container">
      <div class="oxd-toast-icon-wrap oxd-toast-icon-wrap--warn"><i class="oxd-icon bi-exclamation-triangle oxd-toast-icon"></i></div>
    </div>
    <div class="oxd-toast-content oxd-toast-content--warn">
      <p class="oxd-text oxd-text--p oxd-text--toast-title oxd-toast-content-text">Test Toast</p>
      <p class="oxd-text oxd-text--p oxd-text--toast-message oxd-toast-content-text"></p>
    </div>
  </div>
  <div class="oxd-toast-close-container">
    <div class="oxd-toast-close oxd-toast-close--warn" role="button"> × </div>
  </div>
</div>
`;
